rm8-server/UICS/doc/man/man3/uics_state.3
.TH "uics_state" 3 "Wed Oct 20 2021" "Version 1.0" "UICS" \" -*- nroff -*-
.ad l
.nh
.SH NAME
uics_state
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBuics_state\fP (\fBinstHandle\fP *handle)"
.br
.ti -1c
.RI "void \fBrun\fP (int sockfd)"
.br
.RI "main function The run method wait for an incoming command\&. Once all the check up is OK (cmd exist, argument are ok,\&.\&.\&.) we check if the requested callback function is defined\&. "
.ti -1c
.RI "void \fBadd_callback\fP (const char *fname, void(*f)(\fBinstHandle\fP *, \fBcmd\fP *))"
.br
.ti -1c
.RI "void \fBexecute\fP (const char *fname)"
.br
.RI "Description method to execute a standard function\&. A standard function accept one argument of type cmd, return void and must be 1st set using the add_config_func method\&. This method will execute the command named fname with *cc argument\&. "
.ti -1c
.RI "void \fBexecuteOnlyOnce\fP (STATE nextState)"
.br
.ti -1c
.RI "void \fBstart_with\fP (const char *fname)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "std::vector< std::string > \fBffname\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 15 of file uics_state\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "void uics_state::add_callback (const char * fname, void(*)(\fBinstHandle\fP *, \fBcmd\fP *) f)"
Add a function callback to the class\&. A standard function accept one argument of type cmd, return void and must be 1st set using the add_config_func method\&. 
.PP
Definition at line 182 of file uics_state\&.cpp\&.
.PP
.nf
188 {
189     
190     functions\&.add_value(fname,f);
191     ffname\&.push_back(std::string(fname));
192 }
.fi
.SS "void uics_state::execute (const char * fname)"

.PP
Description method to execute a standard function\&. A standard function accept one argument of type cmd, return void and must be 1st set using the add_config_func method\&. This method will execute the command named fname with *cc argument\&. 
.PP
Definition at line 82 of file uics_state\&.cpp\&.
.PP
.nf
91 {
92     void (*f)(instHandle*,cmd*);
93     f = functions[fname];
94     if (f!=nullptr)
95     {
96         f(myHandle,cmd2exec);
97     }
98 
99     return ;
100 }
.fi
.SS "void uics_state::executeOnlyOnce (STATE nextState)"
If this fucntion is called, the \fBuics_state::run\fP will only execute one function\&. This functionallity is usefull for the acquisition combine with a idle state\&. The \fBuics_state(acq)\fP only has 1 function; acquisition()\&. Once the function return, we may tell the \fBuics_state\fP to go to another state e\&.g\&., the idle state 
.PP
Definition at line 127 of file uics_state\&.cpp\&.
.PP
.nf
137 {
138     eOO = true;
139     eOOnextState = nextState;
140 }
.fi
.SS "void uics_state::run (int sockfd)"

.PP
main function The run method wait for an incoming command\&. Once all the check up is OK (cmd exist, argument are ok,\&.\&.\&.) we check if the requested callback function is defined\&. 
.PP
Definition at line 37 of file uics_state\&.cpp\&.
.PP
.nf
44 {
45     while(1)
46     {
47         //if start_with has been used, start by launching this function
48         if (SW)
49         {
50             execute(SW_fname\&.c_str());
51         }
52         int out = getNextConnection(sockfd);
53 
54         if (out==0)
55         {
56             if (!cmd2exec)
57             {
58                 //std::cout<<"super problem!!"<<std::endl;
59                 log_state_transition("Major issue\&.");
60             }
61             else {
62                 if (isInList(cmd2exec->name))
63                 {
64                     execute(cmd2exec->name\&.c_str());
65                     if (eOO)
66                     {
67                         log_state_transition("Transition " +std::to_string(myHandle->nextState)+"->"+std::to_string(eOOnextState));
68                         myHandle->nextState = eOOnextState;
69                     }
70                     break;
71                 }
72                 else
73                 {
74                     log_state_transition("function not recognized");
75                     sndMsg(cmd2exec->sockfd,"function not recognized",nidcsCMD_ERR_NOT_SUPPORTED);
76                 }
77             }
78         }
79         else if (out==EXIT) {break;}
80     }
81 }
.fi


.SH "Author"
.PP 
Generated automatically by Doxygen for UICS from the source code\&.


nutec-server/UICS/src/uics_state.cpp
#include "uics_state.h"
#include "uics_sockets.h"


#define FAILED "failed"
#define SUCCESS "ready"

#define EXIT 229


uics_state::uics_state(instHandle *handle)
{
    myHandle = handle;
    eOO = false;
    SW = false;
    SW_fname = "";
    eOOnextState = myHandle->state;
}

bool uics_state::isInList(std::string name)
/*!
  *Check if name if in function list.
  * */
{
    for (auto n : ffname)
    {   
        if (n.compare(name)==0){return true;}
    }
    return false;
}

void uics_state::start_with(const char *fname)
{
    SW_fname = fname;
    SW = true;
}
void uics_state::run(int sockfd)
/*!
  *\brief main function
  * The run method wait for an incoming command. Once all the check
  * up is OK (cmd exist, argument are ok,...) we check if the requested
  * callback function is defined.
  * */
{
    while(1)
    {
        //if start_with has been used, start by launching this function
        if (SW)
        {
            execute(SW_fname.c_str());
        }
        int out = getNextConnection(sockfd);

        if (out==0)
        {
            if (!cmd2exec)
            {
                //std::cout<<"super problem!!"<<std::endl;
                log_state_transition("Major issue.");
            }
            else {
                if (isInList(cmd2exec->name))
                {
                    execute(cmd2exec->name.c_str());
                    if (eOO)
                    {
                        log_state_transition("Transition " +std::to_string(myHandle->nextState)+"->"+std::to_string(eOOnextState));
                        myHandle->nextState = eOOnextState;
                    }
                    break;
                }
                else
                {
                    log_state_transition("function not recognized");
                    sndMsg(cmd2exec->sockfd,"function not recognized",uicsCMD_ERR_NOT_SUPPORTED);
                }
            }
        }
        else if (out==EXIT) {break;}
    }
}
void uics_state::execute(const char *fname)
/*!
  * \brief Description
  * method to execute a standard function. A standard function
  * accept one argument of type cmd, return void and must be
  * 1st set using the add_config_func method.
  * This method will execute the command named fname with *cc
  * argument.
  */
{
    void (*f)(instHandle*,cmd*);
    f = functions[fname];
    if (f!=nullptr)
    {
        f(myHandle,cmd2exec);
    }

    return ;
}

void uics_state::setNextState(int state)
/*!
  * Will set the handle->nextState to state value.
  * This is required when a change of state is
  * requested since the state handler must know
  * which state to run.
  * */
{
    if (ACQ==state)
    {
        myHandle->nextState = ACQ;
    }
    else if (IDLE==state)
    {
        myHandle->nextState = IDLE;
    }
    else if (INIT==state)
    {
        myHandle->nextState = INIT;
    }
    else if (CONFIG==state)
    {
        myHandle->nextState = CONFIG;
    }
}
void uics_state::executeOnlyOnce(STATE nextState)
/*!
  * If this fucntion is called, the uics_state::run
  * will only execute one function. This functionallity
  * is usefull for the acquisition combine with a
  * idle state. The uics_state(acq) only has 1 function;
  * acquisition(). Once the function return, we may
  * tell the uics_state to go to another state e.g.,
  * the idle state
  */
{
    eOO = true;
    eOOnextState = nextState;
}
int uics_state::getNextConnection(int fd)
/*!
  * \brief Description
  * method that manage incoming commands. Will return
  * -1 if something goes wrong.
  * */
{
    std::string msg="";
    read_socket(&msg,fd);
    if (msg.compare("INCOMINGCMD")==0){
        write_socket(std::string(SUCCESS),fd);

        // Receiving the command
        cmd2exec = new cmd();
        cmd2exec->recvCMD(fd);

        if (!cmd2exec)
        {
            write_socket(std::string(FAILED),fd);
            std::cout<<"big big problem"<<std::endl;
            return -1;
        }

    
    //if fail write on socket -1 else return 0 and write socket
    write_socket(std::string(SUCCESS),fd);
    return 0;
    }
    else if (msg.substr(0,4).compare("EXIT")==0) {
        //request a change of state
        int nexState = std::atoi(msg.substr(5,1).c_str());
        setNextState(nexState);//setting nex state
        write_socket(std::string(SUCCESS),fd);
        return EXIT;
    }
    else {
        write_socket(std::string(FAILED),fd);
        return -1;
    }
}

void uics_state::add_callback(const char *fname, void (*f)(instHandle *, cmd *))
/*!
  * Add a function callback to the class. A standard function
  * accept one argument of type cmd, return void and must be
  * 1st set using the add_config_func method.
  */
{
    
    functions.add_value(fname,f);
    ffname.push_back(std::string(fname));
}





